Scrape categories and topics for `news.py`

tianyizheng02 opened the pull request:
Fixes #204

Replace existing hard-coded news categories and topics with helper functions that scrape these values. The values are scraped when the user calls a function from the news module for the first time, and the scraped values are reused in all subsequent function calls.

tianyizheng02 committed:
Scrape categories and topics for news.py

Replace existing hard-coded news categories and topics with helper
functions that scrape these values. The values are scraped when the user
calls a function from the news module for the first time, and the
scraped values are reused in all subsequent function calls.

azharichenko commented on pittapi/news.py:
```diff
@@ -107,13 +110,39 @@ def _get_page_articles(
     return page_articles
 
 
+def get_categories() -> list[str]:
+    global CATEGORY_URL_NAME_MAP
+    if not CATEGORY_URL_NAME_MAP:
+        CATEGORY_URL_NAME_MAP = _scrape_categories()
+    return list(CATEGORY_URL_NAME_MAP.keys())
+
+
+def get_topics() -> list[str]:
+    global TOPIC_ID_MAP
+    if not TOPIC_ID_MAP:
+        TOPIC_ID_MAP = _scrape_topics()
+    return list(TOPIC_ID_MAP.keys())
```
Fun suggestion, use `functools.cache`!

```suggestion
from functools import cache

@cache
def get_categories() -> list[str]:
    category_map = _scrape_categories()
    return list(category_map.keys())

@cache
def get_topics() -> list[str]:
    topic_id_map = _scrape_topics()
    return list(topic_id_map.keys())
```

tianyizheng02 committed:
Cache categories and topics using @cache

Rewrite news.py to cache categories and topics using the functools
@cache decorator rather than global variables

RitwikGupta commented on Pipfile.lock:
Are these version changes needed? You bump up the minimum Python version

tianyizheng02 replied:
Most of them aren't necessary, but the `jinja2` version change is necessary as a security update (the repo current has 4 Dependabot alerts for security vulnerabilities in the current `jinja2` version). I had simply run `pipenv update` when I was making my changes for this PR.

RitwikGupta replied:
Make the dependency changes in a different PR

RitwikGupta commented on tests/news_test.py:
```diff
@@ -39,16 +47,77 @@ def __init__(self, *args, **kwargs):
+
+        topics = news.get_topics()
+
+        self.assertCountEqual(
```
Is this test going to hold true even into the future?

tianyizheng02 replied:
Yes, all the tests for the `news` module are mocked. The mock files are in the `tests/samples` directory.

tianyizheng02 committed:
Revert dependency updates

Revert dependency updates in Pipfile.lock and requirements.txt. The
updates will be left for a separate PR.
